presto-spi/src/main/java/io/prestosql/spi/block/LongArrayBlockEncoding.java

```diff
@@ -63,24 +65,44 @@ public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceIn
     {
         int positionCount = sliceInput.readInt();
 
-        boolean[] valueIsNull = decodeNullBits(sliceInput, positionCount).orElse(null);
-
+        byte[] valueIsNullPacked = retrieveNullBits(sliceInput, positionCount);
         long[] values = new long[positionCount];
-        if (valueIsNull == null) {
+
+        if (valueIsNullPacked == null) {
             sliceInput.readBytes(Slices.wrappedLongArray(values));
+            return new LongArrayBlock(0, positionCount, null, values);
         }
-        else {
-            int nonNullPositionCount = sliceInput.readInt();
-            sliceInput.readBytes(Slices.wrappedLongArray(values, 0, nonNullPositionCount));
-            int position = nonNullPositionCount - 1;
-            for (int i = positionCount - 1; i >= 0 && position >= 0; i--) {
-                values[i] = values[position];
-                if (!valueIsNull[i]) {
-                    position--;
-                }
+        boolean[] valueIsNull = decodeNullBits(valueIsNullPacked, positionCount);
+
+        int nonNullPositionCount = sliceInput.readInt();
+        sliceInput.readBytes(Slices.wrappedLongArray(values, 0, nonNullPositionCount));
+        int position = nonNullPositionCount - 1;
+
+        // Handle Last (positionCount % 8) values
+        for (int i = positionCount - 1; i >= (positionCount & ~0b111) && position >= 0; i--) {
+            values[i] = values[position];
+            if (!valueIsNull[i]) {
+                position--;
             }
         }
 
+        // Handle the remaining positions.
```

Comment:
rewind `position -= 8` so that both `i` and `position` always point at start of block

Reply:
I don't think this is such a good idea. While `i` is always decremented by 8, `position` always points to the exact index and is being decremented either by 8 or by 1.